Allow configure services for individual attachments

DmitryTsepelev opened the pull request:
_Idea is taken from [this](https://github.com/rails/rails/issues/31419#issuecomment-443549188) comment_

This change allows to configure custom services for individual attachments in the following way:

```ruby
class User < ActiveRecord::Base
  has_one_attached :avatar, service: :s3
end
```

If specified service is not properly configured - `has_one_attached` would throw an error. Since `service_name` is stored inside `active_storage_blobs` table it's also possible that misconfiguration will appear in the realtime - for such cases, there is a special configuration option `Rails.application.configuration.active_storage.service_misconfiguration_behavior`, which can take the following values: `:log` (default), `:fallback_to_default` and `:raise`.

In order to migrate existing apps users would have to re-run `rails active_storage:install` and `rails db:migrate` to add a new migration, suggestions are welcome :)

On the side note, I've moved `require "database/setup"` to the beginning of the test helper to make it look more like a real boot process - migrations are executed before app starts.

DmitryTsepelev committed:
Allow configure services for individual attachments

palkan commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -230,6 +234,31 @@ def purge_later
+        configs = Rails.configuration.active_storage.service_configurations
+
+        begin
+          ActiveStorage::Service.configure service_name, configs
```
This would build a new service instance for each blob–we need some kind of memoization here (`ServiceRegistry.fetch`?). That could also allow us to move exception handling logic there and make `#service` method much more readable:

```ruby
def service
  ActiveStorage::ServiceRegistry.fetch(service_name, self.class.service)
end
```

DmitryTsepelev replied:
I've added memoization, but it looks hard to move exception handling to the same place cause error messages are different enough

palkan reviewed (commented):
> In order to migrate existing apps users would have to re-run rails active_storage:install and rails db:migrate to add a new migration, suggestions are welcome :)

Can we make the presence of `service_name` column optional?

Right now we check for it in the macroses (https://github.com/rails/rails/pull/34935/files#diff-fc054319f93d887d4427a7a728670ba9R159), but we still rely on `service_name` column in `Blob#service`.

What about extracting this _feature check_ (say, `ActiveStorage.per_blob_service_supported?`) and re-use in `Blob#service` to avoid calling `#service_name`?

bogdanvlviv commented on activestorage/db/migrate/20190112182829_add_service_name_to_active_storage_blobs.rb:
```diff
@@ -0,0 +1,5 @@
+  def change
+    add_column :active_storage_blobs, :service_name, :string
+  end
+end
```
> In order to migrate existing apps users would have to re-run rails active_storage:install and rails db:migrate to add a new migration, suggestions are welcome :)

Hey, we've created approach to upgrade apps of users that use Active Storage. See https://github.com/rails/rails/pull/33419.

Could you move this migration under `activestorage/db/update_migrate/` directory so `rake app:update` will copy this migration to existing apps if needed. Also, you should extend `activestorage/db/migrate/20170806125915_create_active_storage_tables.rb` migration to add `service_name` column for new apps on `rails active_storage:install`.

DmitryTsepelev replied:
@bogdanvlviv Done, thanks for the suggestion!

What would happen if user won't execute `rake app:update` before running the app? In case he gets an error preventing him from running the app - we could also drop all [these](https://github.com/rails/rails/pull/34935/files#diff-55562e4b2a8eb648cacecb9d9d4c3ea6R67) checks. Also, [as far as I know](https://github.com/rails/rails/pull/33419#issuecomment-412327595), this task is not designed to be executed in point releases, so we should make this PR a part of 6.0.0 😅

bogdanvlviv replied:
> What would happen if user won't execute `rake app:update` before running the app?

It is ok when an app fails because it hasn't been upgraded well. It is users' responsibility to upgrade it right, our responsibility is to guide on how to do it.

>  In case he gets an error preventing him from running the app - we could also drop all these checks. 

Yeah, I think we can drop these checks.

> Also, as far as I know, this task is not designed to be executed in point releases, so we should make this PR a part of 6.0.0 :sweat_smile:

Right, `rake app:update` is designed only fo major and minor releases. Since this PR brings a new feature, we can release it only in any next major/minor release because we [don't add features in point/patch releases](https://edgeguides.rubyonrails.org/maintenance_policy.html).

DmitryTsepelev replied:
Perfect, it made things way more simpler

bogdanvlviv commented on guides/source/active_storage_overview.md:
```diff
@@ -273,6 +273,25 @@ Call `avatar.attached?` to determine whether a particular user has an avatar:
+`RuntimeError`.
+
+NOTE: if you've initialized ActiveStorage before this feature was shipped, Rails would
+ask you to generate and execute one more migration (`rails active_storage:install && rails db:migrate`).
```
We can remove this `NOTE` when follow this approach https://github.com/rails/rails/pull/34935/files#r248751180

DmitryTsepelev replied:
✔️

bogdanvlviv commented on activestorage/lib/active_storage/engine.rb:
```diff
@@ -93,7 +94,7 @@ class Engine < Rails::Engine # :nodoc:
       ActiveSupport.on_load(:active_storage_blob) do
         if config_choice = Rails.configuration.active_storage.service
-          configs = Rails.configuration.active_storage.service_configurations ||= begin
+          Rails.configuration.active_storage.service_configurations ||= begin
```
Hm, I'm getting error:
```
NoMethodError (Cannot configure service :local for Post#avatar`:
undefined method `deep_symbolize_keys' for nil:NilClass):
```
We should ensure that this block is always executed before `ActiveStorage::ServiceRegistry.fetch(service_name)`  call

DmitryTsepelev replied:
Thanks for catching that! Moved config reading to the ServiceRegistry

DmitryTsepelev commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -230,6 +234,29 @@ def purge_later
+        rescue => e
+          msg = "Cannot configure service :#{service_name} for #{inspect}:"
+
+          case ActiveStorage.service_misconfiguration_behavior
```
Let's imagine a following situation:

There are two services configured in the `yaml` - `:s3_public` and `:s3_private`, the first one is configured as a global default. App records some blobs with a service `:s3_private` and one day someone removes it from the config. In this case we have some blobs saved to the storage, which is not available anymore.

I think it would be helpful to notify a developer that the app is trying to access such a "broken" blob.

DmitryTsepelev commented on activestorage/db/update_migrate/20190112182829_add_service_name_to_active_storage_blobs.rb:
```diff
@@ -0,0 +1,7 @@
+class AddServiceNameToActiveStorageBlobs < ActiveRecord::Migration[6.0]
+  def up
+    unless column_exitsts?(:active_storage_blobs, :service_name)
```
yah, my bad

DmitryTsepelev commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -51,14 +51,16 @@ def find_signed(id)
     # When providing a content type, pass <tt>identify: false</tt> to bypass automatic content type inference.
-    def build_after_upload(io:, filename:, content_type: nil, metadata: nil, identify: true)
-      new(filename: filename, content_type: content_type, metadata: metadata).tap do |blob|
+    # Option <tt>:service</tt> should be provided when blob should be stored in the service different from the
```
✔️

DmitryTsepelev commented:
r? @georgeclaghorn

abhchand commented:
Hey all, just came across this PR while browsing other `activestorage` PRs. 

I just recently opened https://github.com/rails/rails/pull/35290 that adds the ability to configure attachments. 

I don't want to get ahead of myself and assume that change will be accepted. But *if* it is accepted I wanted to suggest that we could use that same mechanism here when configuring the service. 

```ruby
class User < ApplicationRecord
  has_one_attached :avatar do |attachment|
    attachment.service :s3
  end
end
```

Just a thought. Thanks!

DmitryTsepelev commented:
Hi @abhchand, thanks for letting me know! I would add it to my PR if yours will be accepted earlier

iggant commented:
There is no activity on this PR, but will be nice to have this feature in Rails 6.0
@DmitryTsepelev  can you please resolve conflicts and maybe there will be any chance to review this again?

DmitryTsepelev commented:
Hi @iggant, thanks for the support, I'd like to have it merged too :) The only conflict is in the changelog so it shouldn't be a blocker for a review. I believe core team members are busy polishing release candidates

iggant commented:
@bogdanvlviv Can you please review this PR again, will be very nice to have this feature in new major Rails release

lordjavac commented:
Lack of this feature is the one thing preventing me from moving to ActiveStorage. I'd love to see this get implemented soon. Thanks for the hard work!

DmitryTsepelev commented:
Hi @georgeclaghorn, is there a chance this PR can fit into 6.0.0? 🙂

georgeclaghorn commented:
I’m afraid it’s not going to make 6.0. Sorry.

DmitryTsepelev commented:
Got it, thanks for the quick response

iggant commented:
@georgeclaghorn @bogdanvlviv  can you please review this feature, it can be settled in rails 6.1

DmitryTsepelev commented:
Rebased the branch against master to make sure tests are passing

rromanchuk commented:
lol single bucket storage for my entire 2019 'web app', not like there might be a bucket with specific workflows triggered by events, but at least we figured it out for avatars, cant wait to mix my avatar photos with my KYC documents, one IAM policy to rule them all   🤦‍♂  Thank you @DmitryTsepelev for trying to push this through.

gmcgibbon commented on activestorage/lib/active_storage/attached/model.rb:
```diff
@@ -32,7 +32,17 @@ module Attached::Model
-      def has_one_attached(name, dependent: :purge_later)
+      #
+      # If you need the attachment to use a service, which differs from the globally configured one
+      # you should pass +:service+ option, for instance:
```
`If you need the attachment to use a service which differs from the globally configured one, pass the +:service+ option. For instance:`

gmcgibbon commented on activestorage/lib/active_storage/attached/model.rb:
```diff
@@ -88,7 +101,17 @@ def #{name}=(attachable)
-      def has_many_attached(name, dependent: :purge_later)
+      #
+      # If you need the attachment to use a service, which differs from the globally configured one
+      # you should pass +:service+ option, for instance:
```
`If you need the attachment to use a service which differs from the globally configured one, pass the +:service+ option. For instance:`

gmcgibbon commented on activestorage/test/models/attached/many_test.rb:
```diff
@@ -578,6 +577,56 @@ def highlights
   end
 
+  test "attaching a new blob from a Hash with a custom service" do
+    old_service = ActiveStorage::Blob.service
```
You _could_ extract this temporary service setting behaviour into a private method that yields a block, like we've done with `append_on_assign` below.

gmcgibbon commented on activestorage/lib/active_storage/attached/model.rb:
```diff
@@ -130,12 +153,26 @@ def purge_later
+      private
+        def validate_service_configuration(association_name, service)
+          return unless service
+
```
What do you think about simply doing this?
```
unless ServiceRegistry.fetch(service)
  raise ArgumentError, "Cannot configure service :#{service} for #{name}##{association_name}"
end
```
Also, I think that would also mean you don't need `return unless service` above.

DmitryTsepelev replied:
Looks like we still need `return unless service` for cases, when `service` is not explicitly configured and we use a default one (default is validated in the [engine](https://github.com/rails/rails/pull/34935/files#diff-fd33c0087b362ffc63058711a96f5245R108))

gmcgibbon commented on guides/source/active_storage_overview.md:
````diff
@@ -329,6 +345,14 @@ Call `images.attached?` to determine whether a particular message has any images
 @message.images.attached?
 ```
 
+Overriding the default service can be done in a similar way it was done for `has_one_attached`:
````
```suggestion
Overriding the default service is done done the same way as `has_one_attached`, by using the `service` option:
```

gmcgibbon commented on guides/source/active_storage_overview.md:
````diff
@@ -287,6 +287,22 @@ Call `avatar.attached?` to determine whether a particular user has an avatar:
 ```
 
+In some cases you might want to override a default service for a specific attachment.
+It can be done in a following way:
````
```suggestion
You can configure specific services per attachment using the `service` option:
```

gmcgibbon commented on guides/source/active_storage_overview.md:
```diff
@@ -287,6 +287,22 @@ Call `avatar.attached?` to determine whether a particular user has an avatar:
+
+Since the provided value would be stored in the DB, there is a possible situation when
+configuration for the required service is missing. There is a special configuration option
+`Rails.application.configuration.active_storage.service_misconfiguration_behavior` to control
```
Let's use `Rails.application.config.*` when mentioning configuration.

gmcgibbon commented on activestorage/test/dummy/storage/ws/le/wslezbqsjp8l388n6qg93uys7wz3:
```diff
@@ -0,0 +1 @@
+STUFF
```
This should probably be ignored / not committed.

DmitryTsepelev commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -234,6 +238,29 @@ def purge_later
+      if service_name
+        begin
+          ActiveStorage::ServiceRegistry.fetch(service_name)
+        rescue => e
```
The plan that `ServiceRegistry` should always rise the exception and not check the configuration, so we could use it in engine and validations, but now all these cases are covered too.

DmitryTsepelev commented on actiontext/test/dummy/db/migrate/20180212164506_create_active_storage_tables.rb:
```diff
@@ -6,6 +6,7 @@ def change
       t.string   :filename,   null: false
       t.string   :content_type
       t.text     :metadata
+      t.string   :service_name
```
Done, the only question is what should we do with the [update migration](https://github.com/rails/rails/pull/34935/files#diff-de6dd58f12dbb1e4cf145ad39b667766R1) – in this case we should provide some default, what should we pick? Default service name coming from `storage.yml`?

gmcgibbon replied:
Ah, yes, good point. I typically fix this with a migration that adds the nullable column, backfills existing records with a default (in this case the default service name), and then changes the null constraint on the column.

DmitryTsepelev replied:
Changed the upgrade migration [to use](https://github.com/rails/rails/pull/34935/files#diff-de6dd58f12dbb1e4cf145ad39b667766R6) value from `Rails.configuration.active_storage.service`

DmitryTsepelev commented:
@gmcgibbon Thanks for the detailed review! I've addressed most of the issues and rebased the PR against the master. I left a couple of questions to review comments

gmcgibbon commented on activestorage/test/test_helper.rb:
```diff
@@ -85,6 +86,19 @@ def extract_metadata_from(blob)
       Rack::Test::UploadedFile.new file_fixture(filename).to_s
     end
+
+    def append_service
```
I think this is better named `with_service` and make the configured `service` injectable via an argument. That was it is clearer in the tests where the service configuration is coming from.

DmitryTsepelev replied:
agreed, changed it (and added couple more similar methods)

gmcgibbon commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -88,8 +92,11 @@ def create_after_upload!(**arguments_for_create_and_upload) #:nodoc:
+    # Option <tt>:service_name</tt> should be provided when blob should be stored in the service different from the
+    # default one.
+    def create_before_direct_upload!(filename:, byte_size:, checksum:, content_type: nil, metadata: nil, service_name: nil, record: nil)
+      service_name ||= ActiveStorage::ServiceRegistry.name_of(service)
```
Would an `after_initialize { self.service_name ||=  ActiveStorage::ServiceRegistry.name_of(service) }` work?

DmitryTsepelev replied:
I realized this case is only possible when alternative service is not configured for the attachment, so we can just take the default one from `Rails.configuration.active_storage.service` and get rid of `#name_of`.

gmcgibbon commented on activestorage/lib/active_storage/engine.rb:
```diff
@@ -102,26 +103,11 @@ class Engine < Rails::Engine # :nodoc:
-              raise e, "Cannot load `Rails.config.active_storage.service`:\n#{e.message}", e.backtrace
-            end
+          unless ActiveStorage::Blob.service
+            raise ArgumentError, "Cannot load `Rails.config.active_storage.service`:\n#{config_choice}"
```
```suggestion
            raise ArgumentError, "Cannot load `Rails.application.config.active_storage.service`:\n#{config_choice}"
```

gmcgibbon commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -50,30 +50,34 @@ def find_signed(id, record: nil)
     # automatic content type inference.
-    def create_and_upload!(io:, filename:, content_type: nil, metadata: nil, identify: true, record: nil)
-      create_after_unfurling!(io: io, filename: filename, content_type: content_type, metadata: metadata, identify: identify).tap do |blob|
+    # Option <tt>:service_name</tt> should be provided when blob should be stored in the service different from the
```
```suggestion
    # Option <tt>:service_name</tt> should be provided when the blob should be stored in a service different from the
```

gmcgibbon commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -88,8 +92,11 @@ def create_after_upload!(**arguments_for_create_and_upload) #:nodoc:
     # the signed ID.
-    def create_before_direct_upload!(filename:, byte_size:, checksum:, content_type: nil, metadata: nil, record: nil)
-      create! filename: filename, byte_size: byte_size, checksum: checksum, content_type: content_type, metadata: metadata
+    # Option <tt>:service_name</tt> should be provided when blob should be stored in the service different from the
```
```suggestion
    # Option <tt>:service_name</tt> should be provided when the blob should be stored in a service different from the
```

gmcgibbon reviewed (commented):
Thanks for your patience @DmitryTsepelev. I think this LGTM after these last few comments are addressed.

DmitryTsepelev commented:
PR is ready for review again 🙂 I've changed `test_helper` a bit to make sure all the tests share same service configurations (which is required for `ServiceRegistry` to work properly)

palkan reviewed (commented):
I like the idea of non-null `service_name` but I'm not sure that we put the logic of using the default service at the right level (in the `Blob` model).

I think, we the following approach would be a bit simpler and would reduce the responsibility of the Blob:
- Make `service_name` required by the `Blob` model (e.g., `validate :service_name, presence: true`)
- Resolve service name at the macros level (`has_one_attached` / `has_many_attached`), i.e., at the boot time not at dynamically for each Blob created. That means replacing `validate_service_configuration` with `resolve_service_configuration`.

This approach has the following benefits:
- No more `after_initialize` callback (less callbacks — better performance)
- Blob doesn't care about global state anymore — it always uses it "local" information
- No more complex _raises_ and _rescues_ related to the service resolution (we can just use `fetch` without custom exceptions).

@DmitryTsepelev @gmcgibbon WDYT?

gmcgibbon reviewed (commented):
> - Make `service_name` required by the `Blob` model (e.g., `validate :service_name, presence: true`)
> - Resolve service name at the macros level (`has_one_attached` / `has_many_attached`), i.e., at the boot time not at dynamically for each Blob created. That means replacing `validate_service_configuration` with `resolve_service_configuration`.

- Yes, if we have a not-null constraint on `service_name` we should have a model validation as well, good catch.
- If we resolved at the macro level we would have to make the `service_name` required in `Blob`'s creation methods. Since methods to create blobs are public API, isn't that a breaking change? The callback may be slower but it lets us maintain our existing contract.

palkan commented:
@gmcgibbon 

> If we resolved at the macro level we would have to make the service_name required in Blob's creation methods. Since methods to create blobs are public API, isn't that a breaking change? The callback may be slower but it lets us maintain our existing contract.

That's a good point. I'm wondering whether creating blobs were made public intentionally 🤔
I'm not sure that it should be public, Blob is more like an internal "thing". If so, we're OK to change the API a bit. Assuming that we also add a validation, it seems to be a valid change.

gmcgibbon commented:
I think it is intentional, there are use-cases where you can use Active Storage without an attachment model (eg. direct uploads). cc @georgeclaghorn do you have a preference on how we resolve service names?

georgeclaghorn commented:
I prefer the way it is now. Blobs should use the default service if you create them manually without specifying otherwise.

DmitryTsepelev commented on activestorage/test/template/image_tag_test.rb:
```diff
@@ -1,7 +1,6 @@
 # frozen_string_literal: true
 
 require "test_helper"
-require "database/setup"
```
reverted

DmitryTsepelev commented:
Tried to address everything that was discussed above. Also, I've added the [validation](https://github.com/rails/rails/pull/34935/commits/35a8436f5842681f17ef009afdcd05907b57ebed#diff-123aa7e5579a68c3dc535779855c51e1R49) to make sure that invalid service is not passed to the `#update` method.

gmcgibbon commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -36,10 +36,24 @@ class ActiveStorage::Blob < ActiveRecord::Base
 
+  validates :service_name, presence: true
+
+  validate do
```
I wonder what the overhead for unchanged service names would be? It might be worth validating only if the attribute changed.

gmcgibbon reviewed (approved):
LGTM, great job @DmitryTsepelev! Can you squash your commits?

DmitryTsepelev commented:
I've squashed commits and added a check `if service_name_changed? && service_name` to the validation to make sure we do not perform the extra check (however, it feels like `ServiceRegistry#fetch` is cheap because hash has a constant access time)

DmitryTsepelev commented:
Rebased again 🙂

gmcgibbon commented on activestorage/test/models/blob_test.rb:
```diff
@@ -255,4 +270,12 @@ def expected_url_for(blob, disposition: :attachment, filename: nil, content_type
       "https://example.com/rails/active_storage/disk/#{ActiveStorage.verifier.generate(key_params, expires_in: 5.minutes, purpose: :blob_key)}/#{filename}?#{query.to_param}"
     end
+
+    def with_io_logger(io)
```
I think this can be removed.

Hareramrai commented:
Hi @DmitryTsepelev, From which version of rails, we can use this?

DmitryTsepelev commented:
Hi @Hareramrai, it's still in the main [CHANGELOG](https://github.com/rails/rails/blob/master/activestorage/CHANGELOG.md) so I guess it will be released as part of 6.1.

dnalbach commented:
@DmitryTsepelev does your merged PR allow for different S3 _bucket names_ to be usable per attachment, or just the toggle of storage types?

DmitryTsepelev commented:
@dnalbach The line `has_one_attached :avatar, service: :s3` forces ActiveStorage to use the service called `s3` from `storage.yml`, so you can use [any options](https://edgeguides.rubyonrails.org/active_storage_overview.html#amazon-s3-service) (bucket, region, etc.) that are available for the specified adapter. For instance, if you have two buckets `avatars` and `cover_photos` you can do the following:

```yml
# storage.yml
s3_avatars:
  service: S3
  bucket: "photos"
  access_key_id: "..."
  secret_access_key: "..."

s3_cover_photos:
  service: S3
  bucket: "cover_photos"
  access_key_id: "..."
  secret_access_key: "..."
```

```ruby
class User < ApplicationRecord
  has_one_attached :avatar, service: :s3_avatars
  has_one_attached :cover_photo, service: :s3_cover_photos
end
```

dnalbach commented:
Thats awsome @DmitryTsepelev , exactly what I was hoping for. This is a BigDeal in the cloud world I live in. Desperately needed. Look forward to seeing your contribution released!

jules2689 commented:
Hi! This is a great feature and I really appreciate the work done on it! I've opted to start using master on my app to take advantage of this new feature.

One thing I did notice is that the ergonomics of the system becomes a bit awkward with testing environments. Originally I had:
```ruby
class Model < ApplicationModel
 has_one_attached :thing, service: :s3
end
```

But my tests started trying to upload to S3 😅 It was a simple fix for now:
```ruby
class Model < ApplicationModel
 has_one_attached :thing, service: (Rails.env.test? ? :local : :s3)
end
```

But I was wondering if you had thought of some better ergonomics for this given that a ternary statement for every override could become quite awkward. Other things I've considered are:
1. Override the `:s3` storage config in the YAML file with if/else ERB blocks (This feels awkward too)
2. `Service` class that sets constants based on environment (e.g. `Service::THING_SERVICE` could point to `:s3` in dev/prod and `:local` in test). This seems ok, but it's not a pattern that I see a ton in rails

Again, great feature - would love to hear if anyone had any thoughts on a better solution here 🙇

DmitryTsepelev commented:
Hi @jules2689, I'm going to recommend you a trick that will work for your specific case only (s3/disk pair), but it will probably will save a lot of time. Have you heard about [minio](https://github.com/minio/minio)? It's an S3–compatible storage, you can run it locally and keep using "S3" both in prod and dev

palkan commented:
> Override the :s3 storage config in the YAML file with if/else ERB blocks (This feels awkward too)

A better option would be to support per-environment configuration in `storage.yml` and move from service-oriented to context-oriented semantics, e.g.:

```yml
# storage.yml
production:
  images:
    service: s3
  documents:
    service: gcloud

test:
  images:
    service: disk
  documents:
    service: disk
```

And then you just have:

```
# Note that name "storage" would match better than "service" here
has_one_attached :avatar, storage: :images
```

IMO, that's the Rails Way of dealing with this problem.

swissonelabs commented:
I am upgrading an existing app to 6.1.0.alpha

Why doesn't   `rails active_storage:install ` ` rails db:migrate` update my tables?

because of that I don't have service_name in my storage_blob and facing errors.

swissonelabs commented:
`rails active_storage:update`
 worked!

ghiculescu commented:
> `rails active_storage:update`
> worked!

Should this be in the upgrade guide? See https://discuss.rubyonrails.org/t/problem-with-activestorage-blob-on-upgrade-from-6-0-3-4-to-6-1

gokninski commented:
> Hi! This is a great feature and I really appreciate the work done on it! I've opted to start using master on my app to take advantage of this new feature.
> 
> One thing I did notice is that the ergonomics of the system becomes a bit awkward with testing environments. Originally I had:
> 
> ```ruby
> class Model < ApplicationModel
>  has_one_attached :thing, service: :s3
> end
> ```
> 
> But my tests started trying to upload to S3 😅 It was a simple fix for now:
> 
> ```ruby
> class Model < ApplicationModel
>  has_one_attached :thing, service: (Rails.env.test? ? :local : :s3)
> end
> ```
> 
> But I was wondering if you had thought of some better ergonomics for this given that a ternary statement for every override could become quite awkward. Other things I've considered are:
> 
> 1. Override the `:s3` storage config in the YAML file with if/else ERB blocks (This feels awkward too)
> 2. `Service` class that sets constants based on environment (e.g. `Service::THING_SERVICE` could point to `:s3` in dev/prod and `:local` in test). This seems ok, but it's not a pattern that I see a ton in rails
> 
> Again, great feature - would love to hear if anyone had any thoughts on a better solution here 🙇

I've done this using custom configuration in the environment config file (https://guides.rubyonrails.org/configuring.html#custom-configuration)
